Lib/asyncio/streams.py

```diff
@@ -81,16 +124,185 @@
         loop = events.get_event_loop()
 
     def factory():
-        reader = StreamReader(limit=limit, loop=loop,
-                              _asyncio_internal=True)
-        protocol = StreamReaderProtocol(reader, client_connected_cb,
-                                        loop=loop,
-                                        _asyncio_internal=True)
+        protocol = _LegacyServerStreamProtocol(limit,
+                                               client_connected_cb,
+                                               loop=loop,
+                                               _asyncio_internal=True)
         return protocol
 
     return await loop.create_server(factory, host, port, **kwds)
 
 
+class _BaseStreamServer:
+    # TODO: add __repr__
+
+    # Design notes.
+    # StreamServer and UnixStreamServer are exposed as FINAL classes, not function
+    # factories.
+    # async with serve(host, port) as server:
+    #      server.start_serving()
+    # looks ugly.
+    # The class doesn't provide API for enumerating connected streams
+    # It can be a subject for improvements in Python 3.9
+
+    def __init__(self, client_connected_cb,
+                 limit=_DEFAULT_LIMIT,
+                 shutdown_timeout=60,
+                 _asyncio_internal=False):
+        if not _asyncio_internal:
+            raise RuntimeError("_ServerStream is a private asyncio class")
+        self._client_connected_cb = client_connected_cb
+        self._limit = limit
+        self._loop = events.get_running_loop()
+        self._low_server = None
+        self._streams = {}
+        self._shutdown_timeout = shutdown_timeout
+
+    async def bind(self):
+        if self._low_server is not None:
+            return
+        self._low_server = await self._bind()
+
+    def is_bound(self):
+        return self._low_server is not None
+
+    def addresses(self):
+        # I don't want to expose plain socket.socket objects as low-level
```

Comment:
```suggestion
        # We don't want to expose plain socket.socket objects as low-level
```

Please don't use "I" :)